pkg/operator/ceph/cluster/mon/spec.go

```diff
@@ -90,22 +104,212 @@ func (c *Cluster) makeDeployment(config *monConfig, hostname string) *extensions
 	}
+	return []v1.Volume{
+		{Name: k8sutil.DataDirVolume, VolumeSource: dataDirSource},
+		k8sutil.ConfigOverrideVolume(),
+		cephconfig.DefaultConfigVolume(),
+	}
+}
 
-	container := c.monContainer(config, c.clusterInfo.FSID)
+// Mounts for init containers running the Rook binary
+func rookMounts() []v1.VolumeMount {
+	return []v1.VolumeMount{
+		{Name: k8sutil.DataDirVolume, MountPath: k8sutil.DataDir},
+		k8sutil.ConfigOverrideMount(),
+		cephconfig.DefaultConfigMount(),
+	}
+}
+
+// Mounts for containers running Ceph binaries
+func cephMounts() []v1.VolumeMount {
+	return []v1.VolumeMount{
+		{Name: k8sutil.DataDirVolume, MountPath: k8sutil.DataDir},
+		// Rook doesn't run in ceph containers, so it doesn't need the config override mounted
+		cephconfig.DefaultConfigMount(),
+	}
+}
+
+/*
+ * Env Vars
+ */
+// Env vars for containers running the Rook binary
+func (c *Cluster) rookEnvVars(config *monConfig) []v1.EnvVar {
+	return []v1.EnvVar{
+		k8sutil.PodIPEnvVar(k8sutil.PrivateIPEnvVar),
+		PublicIPEnvVar(config.PublicIP),
+		ClusterNameEnvVar(c.Namespace),
+		EndpointEnvVar(),
+		SecretEnvVar(),
+		AdminSecretEnvVar(),
+		k8sutil.ConfigOverrideEnvVar(),
+	}
+}
+
+// Rook doesn't run in daemon containers, so it doesn't need env vars
+
+/*
+ * Network Ports
+ */
+func daemonPorts(config *monConfig) []v1.ContainerPort {
+	return []v1.ContainerPort{
+		{
+			Name:          "client",
+			ContainerPort: config.Port,
+			Protocol:      v1.ProtocolTCP,
+		},
+	}
+}
+
+/*
+ * Security Context
+ */
+// Init and daemon containers share the same context, so we call it 'pod' context
+func podSecurityContext() *v1.SecurityContext {
+	privileged := false
+	if os.Getenv("ROOK_HOSTPATH_REQUIRES_PRIVILEGED") == "true" {
+		privileged = true
+	}
+	return &v1.SecurityContext{Privileged: &privileged}
+}
+
+func (c *Cluster) makeMonPod(config *monConfig, hostname string) *v1.Pod {
+	// Add mons w/ monmaptool w/ args: [--add <mon-name> <mon-endpoint>]...
+	monmapAddMonArgs := []string{}
+	for _, mon := range c.clusterInfo.Monitors {
+		monmapAddMonArgs = append(monmapAddMonArgs, "--add", mon.Name, mon.Endpoint)
+	}
+	monmapFilePath := path.Join(
+		mondaemon.GetMonRunDirPath(c.context.ConfigDir, config.DaemonName),
+		monmapFile,
+	)
+
+	monCommonArgs := []string{ // args needed for all ceph-mon calls
+		"--name", fmt.Sprintf("mon.%s", config.DaemonName),
+		"--cluster", c.clusterInfo.Name,
+		"--mon-data", mondaemon.GetMonDataDirPath(c.context.ConfigDir, config.DaemonName),
+		// It's safe to use the default path for the config b/c 'GenerateConfigFile'
+		// also writes a copy of the config to the default path
+		"--conf", cephconfig.DefaultConfigFilePath(),
+		// It's similarly safe to use the default path for the keyring b/c
+		// 'CreateKeyring' also writes a copy of the keyring to the default path
+		"--keyring", cephconfig.DefaultKeyringFilePath(),
+	}
+
+	// container := c.monContainer(config, c.clusterInfo.FSID)
 	podSpec := v1.PodSpec{
```

Comment:
I tried to copy the way the podSpec is set up from the osd. It has resulted in a fairly long pod spec; I'm not sure how you feel about that, @travisn. I think that despite its length, I still prefer it over setting up the containers in separate functions a la the exploratory mgr work I did here https://github.com/rook/rook/pull/2048/files#diff-b0d5f3a5347e173debf150a8677e1248R315.

Reply:
Separate functions would look better to me too. It definitely crossed my mind when working on the osd pod spec.